packages/block-editor/src/components/use-block-drop-zone/index.native.js

```diff
@@ -0,0 +1,161 @@
+ *
+ * @param {Object}                 blocksLayouts Blocks layouts object.
+ * @param {WPPoint}                position      The position of the item being dragged.
+ * @param {WPBlockListOrientation} orientation   The orientation of a block list.
+ * @param {boolean}                isRTL         Check if current locale is RTL.
+ *
+ * @return {number|undefined} The block index that's closest to the drag position.
+ */
+export function getNearestBlockIndex(
+	blocksLayouts,
+	position,
+	orientation,
+	isRTL
+) {
+	const allowedEdges =
+		orientation === 'horizontal'
+			? [ 'left', 'right' ]
+			: [ 'top', 'bottom' ];
+
+	const isRightToLeft = isRTL;
+
+	let candidateIndex;
+	let candidateDistance;
+
+	// Only enabled for root level blocks.
+	blocksLayouts.forEach( ( element, index ) => {
+		const { x, y, width, height } = element;
+		const rect = {
+			x: element.x,
+			y: element.y,
+			top: y,
+			right: x + width,
+			bottom: y + height,
+			left: x,
+			width,
+			height,
+		};
+		const [ distance, edge ] = getDistanceToNearestEdge(
+			position,
+			rect,
+			allowedEdges
+		);
+
+		if ( candidateDistance === undefined || distance < candidateDistance ) {
+			// If the user is dropping to the trailing edge of the block
+			// add 1 to the index to represent dragging after.
+			// Take RTL languages into account where the left edge is
+			// the trailing edge.
+			const isTrailingEdge =
+				edge === 'bottom' ||
+				( ! isRightToLeft && edge === 'right' ) ||
+				( isRightToLeft && edge === 'left' );
+			const offset = isTrailingEdge ? 1 : 0;
+
+			// Update the currently known best candidate.
+			candidateDistance = distance;
+			candidateIndex = index + offset;
+		}
+	} );
+	return candidateIndex;
+}
+
+/**
+ * @typedef  {Object} WPBlockDropZoneConfig
+ * @property {string} rootClientId The root client id for the block list.
+ */
+
+/**
+ * A React hook that can be used to make a block list handle drag and drop.
+ *
+ * @param {WPBlockDropZoneConfig} dropZoneConfig configuration data for the drop zone.
+ *
+ * @return {Object} An object that contains `targetBlockIndex` and the event
+ * handlers `onBlockDragOver` and `onBlockDragEnd`.
+ */
+export default function useBlockDropZone( {
+	// An undefined value represents a top-level block. Default to an empty
+	// string for this so that `targetRootClientId` can be easily compared to
+	// values returned by the `getRootBlockClientId` selector, which also uses
+	// an empty string to represent top-level blocks.
+	rootClientId: targetRootClientId = '',
+} = {} ) {
+	const targetBlockIndex = useSharedValue( null );
+
+	const { getBlockListSettings, getSettings } = useSelect( blockEditorStore );
+	const {
+		blocksLayouts,
+		getBlockLayoutsOrderedByYCoord,
+	} = useBlockListContext();
+
+	const getSortedBlocksLayouts = useCallback( () => {
+		return getBlockLayoutsOrderedByYCoord( blocksLayouts.current );
+	}, [ blocksLayouts.current ] );
+
+	const isRTL = getSettings().isRTL;
```

Comment:
I think RTL won't affect this logic, as we only support root level blocks for now, which have a vertical layout. However, I'd like to leave here a note as a future reference when we test nested level blocks.

Reply:
Cool! Let's leave it as a reminder that we need to test it once we extend the functionality.